B2g picl merge

shane-tomlinson opened the pull request:
DO NOT MERGE! This branch is to do review and generate discussion.

zaach committed:
include internal api in communication_iframe scripts

zaach committed:
Expose internal watch implementation

zaach committed:
Prepare internalWatch to receive the proper options

seanmonstar committed:
remove persona styling. its been greyed.

Conflicts:
	resources/static/dialog/css/style.css

zaach committed:
We need to set the origin here

seanmonstar committed:
pass event through when using DOMModule event binding

such as this.click(selector, function(e) {});

this lets us access the target, since we can be doing delegation

seanmonstar committed:
wip b2g tospp iframe stuff

zaach committed:
Do a strict check on null-ness of loggedInUser

seanmonstar committed:
use a sandboxed iframe

seanmonstar committed:
allow /tos and /privacy to allow x-frame

Conflicts:
	lib/static/views.js

zaach committed:
network.init() needed for auth check

zaach committed:
failure callback is triggered unless network.clearContext is called after init

seanmonstar committed:
show tos and pp in a modal iframe

seanmonstar committed:
cherry-pick 2472cd8 (adding FxAccounts logo to b2g dialog)

original fixed #2671, but cherry-pick breaks this

Conflicts:
	resources/static/dialog/css/style.css

ozten committed:
Adding forceAuthentication parameter to request

seanmonstar committed:
setting b2g session length to 10 years

effectively gives us forever sessions

fixes #2681

seanmonstar committed:
Revert "pass event through when using DOMModule event binding"

This reverts commit b227d89a1087924352b0f0ed31284ef9523520bb.

jedp committed:
Send internal parameter {silent:true} for silent assertions

seanmonstar committed:
fix x-frame headers tests for /tos and /privacy

ozten committed:
Making forceIssuer optional on address_info and cert_key. Extracted withType.

Rewrote patch to work with latest primary-transition-ux patch on b2g tip.

Conflicts:
	resources/static/dialog/views/set_password.ejs

ozten committed:
Missed adding new dbutils.js

seanmonstar committed:
add gradients back in
(cherry-picked commit cadbdb9e175cbe9a5764e80347dfe6f98c28e613)

Conflicts:
	resources/static/dialog/css/style.css

seanmonstar committed:
fixing cef_logger.alert, and INSERT INTO user
(cherry-picked from 98f9e06930531b6199f7edd42870c5eba4eec994)

ozten committed:
I think a recent merge removed these forceIssuer pieces
(Cherry-picked commit 56624d2320235b5418f18856b807c1f4de52f79f)

seanmonstar committed:
email inserted in mysql should be `unverified`
(Cherry-picked commit 270aad40591bfb659ab24ab4dfa4db0f73a86462)

jedp committed:
Bug 823736 - objects must be stringified between gecko and user js
(Cherry-picked a5eade72a6451e4861cf7f6270085fe2f141effc)

ozten committed:
The verifier service should be explicit about unverified email in its output
(cherry-picked commit a81a55b9d8e323f6fcad25ecd966d1d8735526fb)

seanmonstar committed:
authenticate with unverified email and allowUnverified

fixes #2902

(Cherry-picked commit 3c32c43e0ef7c0a4543237ac320b088f47f22cf2)
Conflicts:
	resources/static/dialog/js/modules/authenticate.js

jedp committed:
Bug 827558 - rename native navigator.id to .mozId

(Cherry-picked commit 5b7004cdc830d28fc78e28ee29517c83216ec408)
Conflicts:
	resources/static/include_js/_jschannel.js

Original changes from include.js applied to _include.js

seanmonstar committed:
send a confirmation email instead of newUser email

since a user is already being created, newUser is the wrong email to
send

(Cherry-picked commit 06f94c3e2f695a9c6274ff4f52f631537a54d198)

seanmonstar committed:
remove unverified certs once an email is verified

fixes #2924

(Cherry-picked commit d308c0d176521124a8cac754279e10a6cb915859)

jedp committed:
Validator should accept FirefoxOS app:// origins
(Cherry-picked commit bf15b527c138ee28817deb8dc979b00775206186)

seanmonstar committed:
mark password reseted emails as verified

if an email is not verified, and the user goes through the
reset password flow, then the email really should be verified.

fixes #2925

(Cherry-picked commit 06721c3483d0c474aa4af6b5c2bad5d5714a4738)

Conflicts:
	lib/db/mysql.js

seanmonstar committed:
removing duplicate emailInfo (wut)
(Cherry-picked commit 73cd47c1ca93cf94a93e06dd00f285fd20aad33e)

seanmonstar committed:
tests for verifying email after password reset

(Cherry-picked commit 649302b4bc8ea8e9fa9612d7ff2f77eb6c63f579)

Conflicts:
	tests/unverified-email-test.js

seanmonstar committed:
make JSON db delete old staged rows
(Cherry-picked commit b9a64140c3c6cbb4fbe99c44f611ee2e8d8dd895)

jedp committed:
Rebased dev and resolved conflicts in these files:

    resources/static/common/js/user.js
    resources/static/dialog/js/misc/helpers.js
    resources/static/dialog/js/modules/authenticate.js
    resources/static/test/cases/common/js/user.js

jedp committed:
An implementation for B2G to ask allow users to log in even if
Persona has not verified that the user owns the email address.

Starts with navigator.id.request({ allowUnverified: true });
When the assertion is returned to the RP, and if they are using
the Persona verifier, must be sure to include an allowUnverified=true
parameter to the verifier, otherwise it will be rejected.

Behind the scenes, when the dialog tries to stage the user, instead
the user is created immediately, an email is added in our email table,
and a stagedEmail token is created and send to the user. When
allowUnverified is true, our wsapi will also check for a staged email/pass,
and use that to allow the user to login.

(merged from commit ebe21f3e3fc9bb80f65f694b359e377ddfad64a0)

jedp committed:
unverified email now adds passwd to user row

this makes things a whole lot simpler underneath

(merged from commit 427052a8f908bc02344f021c85e60c19675fef6e)

shane-tomlinson commented on bin/keysigner:
```diff
@@ -90,14 +92,24 @@ try {
+app.post('/wsapi/cert_key', validate({ 'email': 'email', 'pubkey': 'pubkey', 'ephemeral': 'boolean', 'unverified': { type: 'boolean', required: false }, 'forceIssuer': {type: 'hostname', required: false}}), function(req, resp) {
   var startTime = new Date();
-  cc.enqueue({
+  var forceIssuer = req.params.forceIssuer;
```
I forget, what happens with forceIssuer?

jedp replied:
forceIssuer disables primary lookup and forces the assertion to be issued by the (whitelisted) issuer.  The product case is for marketplace not to have to kick their users over to third-party auth screens in the payment flow.

shane-tomlinson commented on bin/verifier:
```diff
@@ -135,13 +147,11 @@ function doVerification(req, resp, next) {
       });
     } else {
-      resp.json({
+      resp.json(_.extend(r.success, {
```
Are there any fields that come from success that we do not want passed to the client? Should there be a filter to prevent any new fields from automatically being passed?

shane-tomlinson commented on config/b2g.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "authentication_duration_ms": 315360000000
```
A comment about this magic number would be awesome.

jedp replied:
Yes.

I'd prefer to make it 1324512000000, myself.  But either way, it doesn't account for leap years.

shane-tomlinson commented on example/rp/index.html:
```diff
@@ -68,15 +68,21 @@
       <input type="checkbox" id="returnTo">
       <label for="returnTo">Supply returnTo</label><br />
-    </li>
+    </li><li>
```
I dig updated example RPs. And nice cleanup of the HTML.

jedp replied:
Agreed.  @seanmonstar is always so good about that.

shane-tomlinson commented on lib/configuration.js:
```diff
@@ -280,6 +280,10 @@ var conf = module.exports = convict({
     format: 'boolean = false'
+  },
+  forcable_issuers: {
+    doc: "Hostnames which this Persona instance will issue identies for. Used in b2g only.",
```
nit - identies should be identities

shane-tomlinson commented on lib/db/json.js:
```diff
@@ -339,69 +352,70 @@ function getAndDeleteRowForSecret(secret, cb) {
-
-            exports.updatePassword(uid, hash, false, function(err) {
-              cb(err || null, o.email, o.existing_user);
+    dbutils.withType(o.email, function (type) {
```
What happens here if I have an email address and have verified it, then you come along and claim to own it to under the unverified email flow?

shane-tomlinson replied:
Is type here needed because an unverified email could be either primary or secondary and we cannot assume it is secondary like we used to assume?

jedp replied:
I can't claim your email address unless I know the password.  If anything, the danger is the other way around: I mistype my email address without realizing, and eventually you end up verifying it and claim the identity I thought I had.

shane-tomlinson commented on lib/db/json.js:
```diff
@@ -339,69 +352,70 @@ function getAndDeleteRowForSecret(secret, cb) {
+      exports.emailKnown(o.email, function(err, known) {
+        function createAccount() {
+          var emailVal = {};
+          emailVal[o.email] = { type: type, verified: true };
```
Here you add a verified: true flag, but not in completeConfirmEmail, is this expected?

jedp replied:
Excellent catch, Shane.  I'm not sure; let's check that.

shane-tomlinson commented on lib/db/json.js:
```diff
@@ -427,6 +441,8 @@ exports.completePasswordReset = function(secret, password, cb) {
         Object.keys(emails).forEach(function(email) {
           if (email !== o.email && emails[email].type === 'secondary') {
             emails[email].verified = false;
```
We had unverified emails before? Will DBs have to be updated to deal with these?

shane-tomlinson replied:
Are all emails in the DB expected to have the verified field? What about emails we started with nearly 2 years ago? If not, would it be better to add an "unverified" flag for new entries so that old email entries are not affected?

shane-tomlinson commented on lib/db/json.js:
```diff
@@ -587,6 +603,25 @@ exports.cancelAccount = function(authenticated_uid, cb) {
   process.nextTick(function() { cb(null); });
 };
 
+exports.createUnverifiedUser = function(email, hash, cb) {
```
Is there an unverified add, or is that taken care of by the stage function?

jedp replied:
I believe that's it - unverified emails are basically just staged emails that haven't been verified yet

shane-tomlinson commented on lib/static/views.js:
```diff
@@ -71,6 +71,12 @@ function isProductionEnvironment() {
+  '/relay',
+  '/tos',
+  '/privacy'
+];
```
Does /about need added to this list?

shane-tomlinson replied:
While we are at it, can we add '/test/index.html' to this list?

jedp replied:
sounds good

shane-tomlinson commented on lib/wsapi/address_info.js:
```diff
@@ -100,48 +104,63 @@ exports.process = function(req, res) {
+      if (!err && r && r.urls) {
+        typeRightNow = r.urls;
+        typeRightNow.type = 'primary';
+        typeRightNow.issuer = r.authoritativeDomain;
```
This override the issuer that comes in from the request, does this interact with primaries as we expect it to? What is the expected interaction for primaries if an issuer is specified in the request?

jedp replied:
When forceIssuer is used, primary lookup is disabled and it falls right over to the issuer we specify.  So the primaries are not involved at all.

shane-tomlinson commented on resources/static/common/js/storage.js:
```diff
@@ -88,6 +102,14 @@ BrowserID.Storage = (function() {
     return {};
   }
 
+  function getForceIssuerEmails(forceIssuer) {
```
I understand why you are doing it this way, is it possible to combine this series of functions with the ones that already exist? What are the downsides to doing so?

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -120,8 +121,9 @@ BrowserID.User = (function() {
     // onComplete and onFailure functions already set up.
-    stagingStrategy(function(staged) {
-      var status = { success: staged };
+    stagingStrategy(function(status) {
```
Do all of the staging strategies now return an object?

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -839,20 +893,41 @@ BrowserID.User = (function() {
+          var emails_to_remove_pair = [_.difference(client_emails, server_emails)];
+          var emails_to_update_pair = [_.intersection(client_emails, server_emails)];
+
+          if (!! User.forceIssuer && 'default' !== User.forceIssuer) {
```
Instead of creating pairs (my head is ready to explode, it's too small to handle this!), would it be easier to run the old code with each possible issuer? Once for the default, once for User.forceIssuer?

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -1157,85 +1239,91 @@ BrowserID.User = (function() {
+        if ('default' === forceIssuer)
+          storedID = storage.getEmail(email);
+        else
+          storedID = storage.getForceIssuerEmail(email, forceIssuer);
```
I feel like there should be a way to just call storage.getEmail(email, forceIssuer) and have it do the right thing.  There are a lot of if/else's running around that could be collapsed.

jedp replied:
Yes, it looks like there's an opportunity to refactor and collapse a bit

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -1157,85 +1239,91 @@ BrowserID.User = (function() {
+            }, 0);
+          }
+          else {
+            // TODO what will the type of forceIssuer email addresses be?
```
That's a very good question. Should it be a secondary since there are no primaries in B2G?

jedp replied:
Yes, I think we should have a forceIssuer summit soon.  There are still a number of open questions we punted on.  I'll put something on the calendar.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -984,7 +1064,7 @@ BrowserID.User = (function() {
      * @param {function} [onFailure] - Called on XHR failure.
      */
-    addressInfo: function(email, onComplete, onFailure) {
+    addressInfo: function(email, issuer, onComplete, onFailure) {
```
If issuer is already stored at User.forceIssuer, do we need to pass issuer here? Can the code be made easier to understand and more transparent by not passing it? Can we set User.forceIssuer and forget about it from there on out and just have all calls to user.js "do the right thing"?

jedp replied:
That would be great.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -1157,85 +1239,91 @@ BrowserID.User = (function() {
-          }, 0);
-        }, onFailure);
-      }
+    getAssertion: function(email, audience, forceIssuer, onComplete, onFailure) {
```
Same comment about forceIssuer here as with addressInfo, can we forgo passing forceIssuer if User.forceIssuer is already set? We end up with two sources of truth so to speak.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -1137,12 +1217,14 @@ BrowserID.User = (function() {
      * @method syncEmailKeypair
      * @param {string} email - Email address.
-     * @param {string} [issuer] - Issuer of keypair.
+     * @param {string} forceIssuer - IdP that should back the assertion.
```
does forceIssuer need to be added to the function signature?

shane-tomlinson commented on resources/static/dialog/js/modules/dialog.js:
```diff
@@ -13,6 +13,9 @@ BrowserID.Modules.Dialog = (function() {
       helpers = bid.Helpers,
       win = window,
       startExternalDependencies = true,
+      TOSPP_SELECTOR = ".tospp a",
```
I think we should start to make a separate module for CSS selectors that can be shared across other modules.

shane-tomlinson commented on resources/static/dialog/js/modules/dialog.js:
```diff
@@ -146,6 +149,43 @@ BrowserID.Modules.Dialog = (function() {
   }
 
 
+  function showTOSPP(e) {
```
The TOS/PP stuff should move out of here and into its own module so that each can be thought about and verified independently.

shane-tomlinson commented on resources/static/dialog/js/modules/dialog.js:
```diff
@@ -171,6 +211,14 @@ BrowserID.Modules.Dialog = (function() {
 
+      // if B2G
+      //this.click(TOSPP_SELECTOR, showTOSPP);
+      $(TOSPP_SELECTOR).live('click', function(e) {
```
Let's use dom.bindEvent or this.bindEvent instead of live events.

shane-tomlinson commented on resources/static/dialog/js/modules/dialog.js:
```diff
@@ -257,6 +305,27 @@ BrowserID.Modules.Dialog = (function() {
+        // and replace fxos.login.persona.org as the issuer of certs
+        if (paramsFromRP.forceIssuer) {
+          // TODO check for valid domain
+          params.forceIssuer = paramsFromRP.forceIssuer;
```
This should be scrubbed if it hasn't been already.

shane-tomlinson commented on resources/static/dialog/js/modules/provision_primary_user.js:
```diff
@@ -56,7 +54,7 @@ BrowserID.Modules.ProvisionPrimaryUser = (function() {
       }
 
-      user.addressInfo(email, function(status) {
+      user.addressInfo(email, 'default', function(status) {
```
'default' on this line is not future proof.

A user who makes it to this screen now should already have User.forceIssuer set to 'default'. If the restriction barring primaries on B2G is ever lifted, we are going to have to remember to change this line. 'default' (I believe) can be safely set to User.forceIssuer.

shane-tomlinson commented on resources/static/dialog/js/modules/required_email.js:
```diff
@@ -32,7 +32,7 @@ BrowserID.Modules.RequiredEmail = (function() {
 
     function getAssertion() {
-      dialogHelpers.getAssertion.call(self, email, callback);
+      dialogHelpers.getAssertion.call(self, email, 'default', callback);
```
'default' again.

shane-tomlinson commented on resources/static/dialog/js/misc/state.js:
```diff
@@ -343,9 +367,18 @@ BrowserID.State = (function() {
 
-      self.email = email;
+      // qunit tests won't have run start state... reinit selfIssuer
+      self.forceIssuer = self.forceIssuer || 'default';
```
Since forceIssuer is stored in user.js, can the local copy be removed and that one used so there is a single source of truth?

shane-tomlinson commented on resources/static/dialog/js/misc/internal_api.js:
```diff
@@ -125,7 +132,8 @@
       if(authenticated) {
         user.setOrigin(origin);
-        user.getAssertion(email, user.getOrigin(), function(assertion) {
+        var forceIssuer = 'default';
```
Does user.forceIssuer need to be set at all?

shane-tomlinson commented on resources/static/dialog/js/misc/state.js:
```diff
@@ -117,6 +117,10 @@ BrowserID.State = (function() {
       self.hostname = info.hostname;
       self.siteName = info.siteName || info.hostname;
       self.siteTOSPP = !!(info.privacyPolicy && info.termsOfService);
+      self.forceIssuer = user.forceIssuer = (!!info.forceIssuer ? info.forceIssuer : 'default');
```
Some QUnit tests around these new flows would be awesome.

shane-tomlinson commented:
This is _MUCH_ cleaner than the first go!

Overall thoughts:

1) Let's get the tests passing.
2) There are some areas of cleanup, especially on the front end that I would like to see.
2a) Use user.forceIssuer is the source of truth for the issuer. Set this early. Remove issuer from user.js function signatures and from being passed around everywhere on the front end.
2b) storage.js could be cleaned up to remove the specialized functions which use issuer.
3) MOAR QUnit tests!
4) MOAR Selenium tests! I would like to see the expected flows codified so that we can flesh out some of the intricacies.
5) Let's break up the features for merge back to dev.

jedp commented:
Awesome.  Much rejoicing.

Once the tests are passing, and b2g and dev functionality is secure, I'll branch this branch and re-deploy b2g so that we can get all the speedups and other goodness that have landed in dev.

For front-end tests, I have some selenium flows for b2g here: https://github.com/jedp/gaia-ui-tests/commit/6d8674d6418f072943d9bead8703c182c64e9c81 I can work on pulling these into the persona selenium tests.

I'll make a list of commits by feature that we can use to make the job of breaking features up more manageable.

@shane-tomlinson thank you so much for your time and thought in this.  Cheers!

jedp committed:
added function call accidentally dropped in commit c0707ddf

jedp committed:
fixed typo in var name

jedp committed:
Fixing errors introduced in merge in commit c0707ddf

jedp commented:
Pardon the craziness -- rebase dev introduced conflicts

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -258,12 +260,50 @@ BrowserID.User = (function() {
    * @param {function} [onFailure] - Called on error.
    */
   function persistEmailKeypair(email, keypair, cert, onComplete, onFailure) {
+    User.addressInfo(email, User.forceIssuer, function(info) {
```
This needs to be looked at to make sure that caching does not bite us. Only one request is made per address, addressInfo caches responses. If an address's info was checked pre-verification, cached with "state=unverified" and then the address is persisted post-verification, email_obj.unverified will be true.

A better way is to call persistEmailKeypair with an "unverified" flag that is guaranteed to be correct.

Commit:
i18n-abide requires named variables to be replaced to end with `s`

Commit:
/wsapi/stage_reset and requestPasswordReset no longer take a password.

Commit:
Fix the tests! Mostly updates to function signatures and adding a couple of addresses to the mocks.

Commit:
Set the issuer when the dialog first starts and forget about it.

Remove all the extra forceIssuer code everywhere.

Commit:
Remove the *ForceIssuer* functions from storage.js and user.js. Once the issuer is set, it should be forgotten about.

seanmonstar commented on .awsbox.json:
```diff
@@ -9,7 +9,7 @@
   ],
   "env": {
-    "CONFIG_FILES": "$HOME/code/config/production.json,$HOME/code/config/aws.json,$HOME/config.json"
+    "CONFIG_FILES": "$HOME/code/config/production.json,$HOME/code/config/aws.json,$HOME/code/config/b2g.json,$HOME/config.json"
```
we want to merge in this b2g.json config file here?

jedp replied:
right, it should become superfluous.  thanks, good catch.

jedp replied:
but it's not immediately obvious to me what the right solution for `authentication_duration_ms` is

jedp committed:
Removed password param in reset password test (see commit 3c04102e)

jedp committed:
add maximum_event_loop_lag to config.json; can help avoid random test failures

Commit:
Put the B2G TOS/PP code into its own module insted of dialog.js

Add some unit tests for it too.

Commit:
Move the TOS/PP HTML into a template.

Commit:
Change the B2G TOS/PP close button from a span to an anchor.

Commit:
Start a stub for fixupIssuer which will scrub the issuer.

Commit:
Fix up rebase and JSHint errors.

Commit:
Move the allowUnverified code from network.js to user.js

Update tests. Add a test for network.js->certKey

Commit:
Add JedP's B2G example RP to the repo. Bound to port 10013

Commit:
Start on the work to get an assertion for the forceIssuer.

Commit:
Fix where include.js comes from.

Commit:
Update helpers.log to handle multiple arguments. Objects will be JSON.stringified.

Commit:
Add a few basic tests for internal.watch.

When loading up the internal_api, only set user.setIssuer if an issuer is saved for the current domain.

Commit:
Add QUnit tests for forceAuthentication in dialog.js->doCheckAuth

Commit:
Update B2G TOS/PP to only take effect if inlineTermsOfService is passed as an option to the dialog.

* Rename B2gTosPp to InlineTosPp

shane-tomlinson commented:
Closing this down because we broke it into its component parts.
